control-plane/agents/src/bin/core/volume/operations.rs

```diff
@@ -964,86 +870,153 @@ impl CreateVolumeExeVal for CreateVolume {
     }
 }
 
+#[derive(Debug, Clone)]
+pub struct ReplicaWithLabels {
+    replicas: Vec<Replica>,
+    #[allow(dead_code)]
+    labels: Option<HashMap<String, String>>,
+}
+
+impl ReplicaWithLabels {
+    pub fn new(replicas: Vec<Replica>, labels: Option<HashMap<String, String>>) -> Self {
+        Self { replicas, labels }
+    }
+
+    #[allow(dead_code)]
+    pub fn replicas(&self) -> Vec<Replica> {
+        self.replicas.clone()
+    }
+}
+
 #[async_trait::async_trait]
 impl CreateVolumeExe for CreateVolume {
-    type Candidates = CreateReplicaCandidate;
+    type Candidates = Vec<CreateReplicaCandidate>;
+    type Replicas = ReplicaWithLabels;
 
-    async fn setup<'a>(
-        &'a self,
-        context: &mut Context<'a>,
-    ) -> Result<CreateReplicaCandidate, SvcError> {
-        // todo: pick nodes and pools using the Node&Pool Topology
+    async fn setup<'a>(&'a self, context: &mut Context<'a>) -> Result<Self::Candidates, SvcError> {
         // todo: virtually increase the pool usage to avoid a race for space with concurrent calls
         create_volume_replicas(context.registry, self, context.volume.as_ref()).await
     }
 
     async fn create<'a>(
         &'a self,
         context: &mut Context<'a>,
-        candidates: CreateReplicaCandidate,
-    ) -> Vec<Replica> {
-        let mut replicas = Vec::<Replica>::with_capacity(candidates.candidates().len());
-        for replica in candidates.candidates() {
-            if replicas.len() >= self.replicas as usize {
-                break;
-            } else if replicas.iter().any(|r| {
-                r.node == replica.node
-                    || spread_label_is_same(r, replica, context).unwrap_or_else(|error| {
+        candidates: Self::Candidates,
+    ) -> Result<Self::Replicas, SvcError> {
+        let mut final_replicas = Vec::<Replica>::new();
+        let mut hashmap: Option<HashMap<String, String>> = Some(HashMap::new());
+        // In case of affinty key, if some replica are already provisioned then delete them.
+        let request = GetSuitablePools::new(context.volume.as_ref(), None);
+        let is_affinity = is_affinity_requested(request.clone());
+
+        if context.volume.as_ref().num_replicas > 0 && is_affinity {
+            match self.undo(context, &final_replicas).await {
+                Ok(_) => {}
+                Err(_) => {
+                    return Err(SvcError::ReplicaCreateNumber {
+                        id: context.volume.uid_str(),
+                    });
+                }
+            }
+        }
+        for create_replica_candidate in candidates.iter() {
+            let mut replicas =
```

Comment:
Why do we need a final_replicas, can't we just reset it on the for loop and just reuse it?
AFAICT you're never modifying final_replicas, you're only assigning it?